src/inet/TCPEndPointImplLwIP.cpp

```diff
@@ -77,32 +71,37 @@ CHIP_ERROR TCPEndPointImplLwIP::BindImpl(IPAddressType addrType, const IPAddress
     {
         if (reuseAddr)
         {
-            ip_set_option(mTCP, SOF_REUSEADDR);
+            RunOnTCPIP([this]() { ip_set_option(mTCP, SOF_REUSEADDR); });
         }
         res = addr.ToLwIPAddr(addrType, ipAddr);
     }
 
     if (res == CHIP_NO_ERROR)
     {
-        res = chip::System::MapErrorLwIP(tcp_bind(mTCP, &ipAddr, port));
+        res = chip::System::MapErrorLwIP(RunOnTCPIPRet([this, &ipAddr, port]() { return tcp_bind(mTCP, &ipAddr, port); }));
     }
 
-    // Unlock LwIP stack
-    UNLOCK_TCPIP_CORE();
     return res;
 }
 
 CHIP_ERROR TCPEndPointImplLwIP::ListenImpl(uint16_t backlog)
 {
-    // Start listening for incoming connections.
-    mTCP              = tcp_listen(mTCP);
     mLwIPEndPointType = LwIPEndPointType::TCP;
-
-    tcp_arg(mTCP, this);
-
-    tcp_accept(mTCP, LwIPHandleIncomingConnection);
-
-    return CHIP_NO_ERROR;
+    CHIP_ERROR err    = CHIP_NO_ERROR;
+    if (!mPreAllocatedConnectEP)
+    {
+        err = GetEndPointManager().NewEndPoint(&mPreAllocatedConnectEP);
```

Comment:
After this call returns successfully, mPreAllocatedConnectEP is pointing to something we retained, right?  Where is the matching release?

Reply:
Added releasing logic in the close function of listen endpoint.
And added documents to describe when the pre-allocated endpoint will be released